btx/processing/pipca.py

```diff
@@ -0,0 +1,1011 @@
+    PsanaInterface,
+    bin_data,
+    bin_pixel_index_map,
+    retrieve_pixel_index_map,
+    assemble_image_stack_batch,
+)
+
+class PiPCA:
+
+    """Parallelized Incremental Principal Component Analysis."""
+
+    def __init__(
+        self,
+        exp,
+        run,
+        det_type,
+        start_offset=0,
+        num_images=10,
+        num_components=10,
+        batch_size=10,
+        priming=False,
+        benchmarking=False,
+        downsample=False,
+        bin_factor=2,
+        output_dir="",
+    ):
+
+        self.comm = MPI.COMM_WORLD
+        self.rank = self.comm.Get_rank()
+        self.size = self.comm.Get_size()
+
+        self.psi = PsanaInterface(exp=exp, run=run, det_type=det_type)
+        self.psi.counter = start_offset
+
+        self.priming = priming
+        self.benchmarking = benchmarking
+        self.downsample = downsample
+        self.bin_factor = bin_factor
+        self.output_dir = output_dir
+
+        (
+            self.num_images,
+            self.num_components,
+            self.batch_size,
+            self.num_features,
+        ) = self.set_params(num_images, num_components, batch_size, bin_factor)
+
+        # compute number of counts in and start indices over ranks
+        self.split_indices, self.split_counts = distribute_indices(
+            self.num_features, self.size
+        )
+
+        # attribute for storing interval data
+        self.task_durations = dict({})
+
+        self.num_incorporated_images = 0
+        self.outliers, self.pc_data = [], []
+
+    def get_params(self):
+        """
+        Method to retrieve iPCA params.
+
+        Returns
+        -------
+        num_incorporated_images : int
+            number of images used to build model
+        num_components : int
+            number of components maintained in model
+        batch_size : int
+            batch size used in model updates
+        num_features : int
+            dimensionality of incorporated images
+        """
+        return (
+            self.num_incorporated_images,
+            self.num_components,
+            self.batch_size,
+            self.num_features,
+        )
+
+    def set_params(self, num_images, num_components, batch_size, bin_factor):
+        """
+        Method to initialize iPCA parameters.
+
+        Parameters
+        ----------
+        num_images : int
+            Desired number of images to incorporate into model.
+        num_components : int
+            Desired number of components for model to maintain.
+        batch_size : int
+            Desired size of image block to be incorporated into model at each update.
+        bin_factor : int
+            Factor to bin data by.
+
+        Returns
+        -------
+        n : int
+            Number of images to incorporate into model.
+        q : int
+            Number of components for model to maintain.
+        m : int
+            Size of image block to be incorporated into model at each update.
```

Comment:
Since this is the same as `batch_size`, would it simplify matters to just have one variable name?

Reply:
I agree. @calhep maybe let's revisit those variable names: `n`, `q`, `m`, `d` to make them more intuitively understandable?

Reply:
Hi @fredericpoitevin , I agree that the switching in convention is confusing, the rationale was to provide a distinction between raw user-supplied parameters and the processed parameters from `set_params` which may differ. I can consolidate the naming convention to be consistent and improve readability.